compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -1049,15 +1049,34 @@ class Typer extends Namer
      */
     var paramIndex = Map[Name, Int]()
 
-     /** If function is of the form
+    /** Infer parameter type from the body of the function
+     *
+     *  1. If function is of the form
+     *
      *      (x1, ..., xN) => f(... x1, ..., XN, ...)
+     *
      *  where each `xi` occurs exactly once in the argument list of `f` (in
      *  any order), the type of `f`, otherwise NoType.
+     *
+     *  2. If the function is of the form
+     *
+     *     (using x1, ..., xN) => f
+     *
+     *  where `f` is a contextual function type of the form `(T1, ..., TN) ?=> T`,
+     *  then `xi` takes the type `Ti`.
+     *
      *  Updates `fnBody` and `paramIndex` as a side effect.
      *  @post: If result exists, `paramIndex` is defined for the name of
      *         every parameter in `params`.
      */
-    lazy val calleeType: Type = fnBody match {
+    lazy val calleeType: Type = untpd.stripAnnotated(fnBody) match {
+      case ident: untpd.Ident if isContextual =>
+        val tp = typedIdent(ident, WildcardType).tpe.widen
+        if defn.isContextFunctionType(tp) && params.size == defn.functionArity(tp)
+        then
```

Comment:
Formatting: `then` should go on the previous line here. Use a newline only after a multi-line condition.